dll/win32/shell32/dialogs/drive.cpp

```diff
@@ -47,18 +65,87 @@ typedef struct
+    wszSystemDrv[0] = UNICODE_NULL;
+    GetSystemDirectory(wszSystemDrv, _countof(wszSystemDrv));
+    return wszSystemDrv[0] == pContext->Drive + L'A';
+}
+
+static HANDLE
+OpenLogicalDriveHandle(WORD DriveNumber)
+{
+    const WCHAR szPath[] = { '\\', '\\', '?', '\\', WCHAR(DriveNumber + 'A'), ':', '\0' };
+    return CreateFileW(szPath, FILE_READ_DATA, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                       NULL, OPEN_EXISTING, 0, NULL);
+}
 
-    wszDriveLetter[0] = pContext->Drive + L'A';
-    StringCchCatW(wszDriveLetter, _countof(wszDriveLetter), L":");
+static BOOL
+GetLogicalDriveSize(WORD DriveNumber, ULARGE_INTEGER &Result)
+{
+    BOOL bSuccess = FALSE;
+    const WCHAR szDrivePath[] = { WCHAR(DriveNumber + 'A'), ':', '\\', '\0' };
 
-    if (!GetEnvironmentVariableW(L"SystemDrive", wszSystemDrv, _countof(wszSystemDrv)))
-        return FALSE;
+    HANDLE hDevice = OpenLogicalDriveHandle(DriveNumber);
+    if (hDevice != INVALID_HANDLE_VALUE)
+    {
+        GET_LENGTH_INFORMATION LengthInfo;
+        bSuccess = DevIoCtl(hDevice, IOCTL_DISK_GET_LENGTH_INFO, NULL, 0, &LengthInfo, sizeof(LengthInfo));
+        Result.QuadPart = LengthInfo.Length.QuadPart;
+        if (!bSuccess && GetDriveTypeW(szDrivePath) == DRIVE_REMOVABLE) // Blank floppy
+        {
+            DISK_GEOMETRY dg;
+            bSuccess = DevIoCtl(hDevice, IOCTL_DISK_GET_DRIVE_GEOMETRY, NULL, 0, &dg, sizeof(dg));
+            if (bSuccess)
+                Result.QuadPart = dg.Cylinders.QuadPart * dg.TracksPerCylinder * dg.SectorsPerTrack * dg.BytesPerSector;
+        }
+        CloseHandle(hDevice);
+    }
+    if (!bSuccess || !Result.QuadPart)
+        bSuccess = GetDiskFreeSpaceExW(szDrivePath, NULL, &Result, NULL); // Note: Not exact if NTFS quotas are in effect
+    return bSuccess;
+}
 
-    if (!_wcsicmp(wszDriveLetter, wszSystemDrv))
-        return TRUE;
+static PWSTR
+CreateTipText(FORMAT_DRIVE_CONTEXT &Ctx)
+{
+    HANDLE hDevice = OpenLogicalDriveHandle(Ctx.Drive);
+    if (hDevice == INVALID_HANDLE_VALUE)
+        return NULL;
+
+    ULONG cb;
+    ZeroMemory(&Ctx.Tip, sizeof(Ctx.Tip));
+    NtQueryObject(hDevice, ObjectNameInformation, &Ctx.Tip, sizeof(Ctx.Tip), &cb);
+    if (Ctx.Tip.Name.Buffer && Ctx.Tip.Name.Buffer[0] == '\\')
+        StringCbCatW(Ctx.Tip.Name.Buffer, sizeof(Ctx.Tip) - sizeof(UNICODE_STRING), L"\n");
+    else
+        (Ctx.Tip.Name.Buffer = Ctx.Tip.Buffer)[0] = UNICODE_NULL;
 
-    return FALSE;
+    PARTITION_INFORMATION_EX pie;
+    if (!DevIoCtl(hDevice, IOCTL_DISK_GET_PARTITION_INFO_EX, NULL, 0, &pie, sizeof(pie)))
+    {
+        pie.PartitionStyle = PARTITION_STYLE_RAW;
+        PARTITION_INFORMATION pi;
+        if (DevIoCtl(hDevice, IOCTL_DISK_GET_PARTITION_INFO, NULL, 0, &pi, sizeof(pi)))
+        {
+            pie.PartitionStyle = PARTITION_STYLE_MBR;
+            pie.PartitionNumber = pi.PartitionNumber;
+        }
+    }
+    CloseHandle(hDevice);C_ASSERT(sizeof(Ctx.Tip) > 800);
```

Comment:
Can move this `C_ASSERT` to somewhere closer to the struct?
Or even make it a `static_assert` next to the struct (since this is c++ code)